services/hierarchy/src/main/kotlin/VulnerabilityService.kt

```diff
@@ -0,0 +1,146 @@
+                (
+                        VulnerabilitiesTable innerJoin
+                                AdvisorResultsVulnerabilitiesTable innerJoin
+                                AdvisorResultsTable innerJoin
+                                AdvisorRunsIdentifiersTable innerJoin
+                                AdvisorRunsTable innerJoin
+                                AdvisorJobsTable innerJoin
+                                IdentifiersTable innerJoin
+                                VulnerabilityReferencesTable
+                        ).select(
+                        VulnerabilitiesTable.id,
+                        VulnerabilitiesTable.externalId,
+                        VulnerabilitiesTable.summary,
+                        VulnerabilitiesTable.description,
+                        IdentifiersTable.type,
+                        IdentifiersTable.namespace,
+                        IdentifiersTable.name,
+                        IdentifiersTable.version,
+                        VulnerabilityReferencesTable.url,
+                        VulnerabilityReferencesTable.scoringSystem,
+                        VulnerabilityReferencesTable.severity
+                    ).where { AdvisorJobsTable.ortRunId eq ortRunId }
+            }
+
+        val result = mutableListOf<VulnerabilityWithIdentifier>()
+        queryResult.data.forEach { vulnerabilityWithIdentifierAndReference ->
+            collectVulnerabilitiesWithIdentifiers(result, vulnerabilityWithIdentifierAndReference)
+        }
+
+        ListQueryResult(result, parameters, result.size.toLong())
+    }
+
+    private fun collectVulnerabilitiesWithIdentifiers(
+        result: MutableList<VulnerabilityWithIdentifier>,
+        vulnsWithIdAndReference: VulnerabilityWithIdentifierAndReference
+    ) {
+        val identifier = vulnsWithIdAndReference.identifier
+        val vulnerability = vulnsWithIdAndReference.vulnerability
+
+        val existingVulnerabilityWithIdentifier =
+            result.find { it.vulnerability.externalId == vulnerability.externalId }
+        if (existingVulnerabilityWithIdentifier != null) {
+            val updatedReferences = existingVulnerabilityWithIdentifier.vulnerability.references.toMutableList()
+            updatedReferences += vulnsWithIdAndReference.reference
+
+            val updatedVulnerability =
+                existingVulnerabilityWithIdentifier.vulnerability.copy(references = updatedReferences)
+
+            result.remove(existingVulnerabilityWithIdentifier)
+            result.add(VulnerabilityWithIdentifier(updatedVulnerability, identifier))
+        } else {
+            result.add(
+                VulnerabilityWithIdentifier(
+                    vulnerability = vulnerability.copy(
+                        references = mutableListOf(vulnsWithIdAndReference.reference)
+                    ),
+                    identifier = identifier
+                )
+            )
+        }
+    }
```

Comment:
Maybe I don't understand the algorithm correctly, but does this not change the size of the result set? For example, if the query has a limit of 20 and the first two vulnerabilities have each 10 references, then only 2 items are returned? And the first vulnerability has more than 20 references, than only the first 20 are returned?

Reply:
Good catch. The implementation was incorrect.
Changed it, and added a test for it.